runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -1330,20 +1334,19 @@ private void process(
       WorkItemCommitRequest commitRequest = outputBuilder.build();
       int byteLimit = maxWorkItemCommitBytes;
       int commitSize = commitRequest.getSerializedSize();
+      int estimatedCommitSize = commitSize < 0 ? Integer.MAX_VALUE : commitSize;
+
       // Detect overflow of integer serialized size or if the byte limit was exceeded.
-      windmillMaxObservedWorkItemCommitBytes.addValue(
-          commitSize < 0 ? Integer.MAX_VALUE : commitSize);
-      if (commitSize < 0) {
-        throw KeyCommitTooLargeException.causedBy(computationId, byteLimit, commitRequest);
-      } else if (commitSize > byteLimit) {
-        // Once supported, we should communicate the desired truncation for the commit to the
-        // streaming engine. For now we report the error but attempt the commit so that it will be
-        // truncated by the streaming engine backend.
+      windmillMaxObservedWorkItemCommitBytes.addValue(estimatedCommitSize);
+      if (estimatedCommitSize > byteLimit) {
         KeyCommitTooLargeException e =
             KeyCommitTooLargeException.causedBy(computationId, byteLimit, commitRequest);
         reportFailure(computationId, workItem, e);
         LOG.error(e.toString());
+
+        commitRequest = buildWorkItemTruncationRequest(key, workItem, estimatedCommitSize);
```

Comment:
Instead of reinitializing everything from scratch, can't you just reuse the existing commitRequest? Something like:

commitRequest = commitRequest.toBuilder(). setExceedsMaxWorkItemCommitBytes(true).setEstimatedWorkItemCommitBytes(estimatedCommitSize).build()

Reply:
All of the 'doomed' messages, timers, counters, etc. are being cleared as well.  It would be a little more efficient to do that explicitly rather than just start over, but less future-proof as new fields are added to the proto and need to be cleared out here as well.

Reply:
Could you add a comment to this effect so that it's clear in the code why you're doing it this way?

Reply:
Sure - just added it and re-squashed the commits